pkg/api/v1/types.go

```diff
@@ -710,6 +725,22 @@ type Container struct {
 	TTY   bool `json:"tty,omitempty" description:"Whether this container should allocate a TTY for itself, also requires 'stdin' to be true; default is false"`
 }
 
+func (ctr *Container) ApplyDefaults() {
```

Comment:
Another wrinkle with declarative defaults in the case of non-literal
values:  field names.  Suppose I want to say that field `Foo` has the same
value as `Bar` by default:

```
type fooBar struct {
    Foo int `json:"foo,omitempty" default:"field(bar)"`
    Bar int `json:"bar" default:"val(93)"`
}
```

simply emitting code like `obj.Foo = obj.Bar` is easy, but... we don't want
to say Bar because the API uses "bar", and we can't say "bar" because Go
uses "Bar".  Now I have to either parse JSON tags and lookup fields in the
struct by the JSON name or else blindly JSON-ize the field name.  I guess
blindly JSON-izing isn't toooo bad.




On Tue, Sep 1, 2015 at 5:57 PM, Tim Hockin <notifications@github.com> wrote:

> In pkg/api/v1/types.go
> <https://github.com/kubernetes/kubernetes/pull/13177#discussion_r38490317>
> :
>
> > @@ -710,6 +725,22 @@ type Container struct {
> >  	TTY   bool `json:"tty,omitempty" description:"Whether this container should allocate a TTY for itself, also requires 'stdin' to be true; default is false"`
> >  }
> >
> > +func (ctr *Container) ApplyDefaults() {
>
> Some of our defaults are "same as that other field". Maybe possible to do
> that declaratively, but I expect corner cases.
> … <#14f8b8f68c62c29d_>
> On Tue, Sep 1, 2015 at 2:52 PM, Brendan Burns <notifications@github.com>
> wrote: In pkg/api/v1/types.go <
> https://github.com/kubernetes/kubernetes/pull/13177#discussion_r38476943>
> : > @@ -710,6 +725,22 @@ type Container struct { > TTY bool
> `json:"tty,omitempty" description:"Whether this container should allocate a
> TTY for itself, also requires 'stdin' to be true; default is false"` > } >
> > +func (ctr *Container) ApplyDefaults() { To be clear: validation in code
> is great and appropriate, tags are probably not a good fit for the reasons
> you mentioned. defaulting in tags seems to make sense, it's
> self-documenting, and I don't think you should ever need code to do it.
> --brendan — Reply to this email directly or view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/13177/files#r38476943>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13177/files#r38490317>.
>